h/h_api/bulk_api/command_processor.py

```diff
@@ -0,0 +1,186 @@
+
+from h.h_api.bulk_api.command_batch import CommandBatch
+from h.h_api.bulk_api.id_references import IdReferences
+from h.h_api.bulk_api.model.command import ConfigCommand
+from h.h_api.bulk_api.model.report import Report
+from h.h_api.bulk_api.observer import Observer
+from h.h_api.enums import CommandStatus
+from h.h_api.exceptions import CommandSequenceError, InvalidDeclarationError
+
+
+class CommandProcessor:
+    """
+    Manager which will check and run a number of bulk API commands.
+
+    The manager is responsible for:
+
+     * Checking the correctness of commands and their order
+     * Batching similar commands together for batch processing
+     * Dispatching these batches to an `Executor` to make changes
+     * Reporting each command to an `Observer` to look at
+    """
+
+    def __init__(self, executor, observer=None, batch_size=100):
+        """
+        :param executor: An executor to carry out commands
+        :param observer: An observer to view commands
+        :param batch_size: Commands to wait for before executing
+        """
+
+        self.executor = executor
+        self.observer = observer or Observer()
+
+        # Pass a reference to the command batch to our execute batch command
+        # so it can call us back when the batch is ready
+        self.batch = CommandBatch(on_flush=self._execute_batch, batch_size=batch_size)
+
+        # A container for any custom references to objects
+        self.id_refs = IdReferences()
+        self.reports = defaultdict(list)
+        self.config = None
+        self.command_count = 0
+
+    def process(self, commands):
+        """Process the given commands.
+
+        :param commands: An iterable of Command objects
+        """
+
+        for command in commands:
+            self._process_single_command(command)
+
+        # Flush out the last batch of commands (if any)
+        self.batch.flush()
+
+        self._check_command_count(final=True)
+
+        self._report_back()
+
+    def _process_single_command(self, command):
+        """Process a single command."""
+
+        self.observer.observe_command(command, status=CommandStatus.AS_RECEIVED)
+
+        self.command_count += 1
+        self._check_command_count()
+
+        if isinstance(command, ConfigCommand):
+            self._configure(command.body)
+        else:
+            self._add_to_batch(command)
+
+        self.observer.observe_command(command, status=CommandStatus.POST_EXECUTE)
+
+    def _configure(self, config):
+        """Configure this object and the executor."""
+
+        if self.config is not None:
+            raise CommandSequenceError("Cannot currently re-configure jobs")
+
+        self.executor.configure(config)
+        self.config = config
+
+    def _add_to_batch(self, command):
+        """Add a single command to the batch.
+
+        When the batch size reaches maximum, or the task switches this will
+        trigger the commands to be flushed to the executor.
```

Comment:
```suggestion
This may cause the CommandBatch to call the on_flush() callback that we passed to it
(self._execute_batch()) if it decides that it's time to execute the next batch.
```